src/features/textEditor/components/tipTapNodes/citation/citationNode.ts

```diff
@@ -0,0 +1,193 @@
+ * Adding text in the citation node actually adds the '; ' text to the document
+ * */
+const citationPlugin = new Plugin<CitationPluginState>({
+  key: citationPluginKey,
+
+  state: {
+    init: () => null,
+    apply: (tr, _value, _oldState, newState) => {
+      if (!newState.selection.empty) {
+        return null;
+      }
+
+      const { $from } = newState.selection;
+
+      // If the transaction does not happen in a citation node or is not at the end of the citation node,
+      // there is nothing to do
+      if ($from.parent.type.name !== citationNode.name || $from.pos !== $from.after() - 1) {
+        return null;
+      }
+
+      const { parent } = $from;
+
+      // If the last child of the citation node is not a reference, nothing to do
+      if (parent.child(parent.childCount - 1).type.name !== referenceNode.name) {
+        return null;
+      }
+
+      // Otherwise, add a decoration
+      return newState.selection.from;
+    },
+  },
+
+  // Inserts '; ' in the document when writing something after a reference, at the end of a citation node
+  appendTransaction: (transactions, oldState, newState) => {
+    if (!newState.selection.empty) {
+      return null;
+    }
+
+    const { tr } = newState;
+    const oldPluginState = citationPluginKey.getState(oldState) as CitationPluginState;
+
+    // If there was no decoration before the update, nothing to do
+    if (!oldPluginState) {
+      return null;
+    }
+
+    const { $from } = newState.selection;
+    const { parent } = $from;
+
+    // If the selection after the update is not at the end of the citation node, nothing to do
+    if (parent.type.name !== citationNode.name || $from.nodeAfter) {
+      return null;
+    }
+
+    // Otherwise, if there was a replace step that added something in the citation node, we insert '; ' in the document
+    if (
+      transactions.find((transaction) =>
+        transaction.steps.find(
+          (step) => step instanceof ReplaceStep && step.from === step.to && step.from === oldPluginState,
+        ),
+      )
+    ) {
+      tr.insertText('; ', oldPluginState);
+      return tr;
+    }
+
+    return null;
+  },
+
+  props: {
+    decorations(state) {
+      const decorationPos = this.getState(state) as CitationPluginState;
+      if (!decorationPos) {
+        return null;
+      }
+      return DecorationSet.create(state.doc, [
+        Decoration.widget(
+          decorationPos,
+          () => {
+            const parentNode = document.createElement('span');
+
+            parentNode.innerHTML = '; ';
+            parentNode.style.color = 'hsl(var(--color-muted))';
+
+            return parentNode;
+          },
+          {
+            side: -1,
+          },
+        ),
+      ]);
+    },
+  },
+});
+
+/**
+ * Plugin that prevents the cursor from being directly after the opening and closing brackets of citation nodes
+ */
+const moveCursorPluginKey = new PluginKey('moveCursor');
+const moveCursorPlugin = new Plugin({
```

Comment:
not directly related to this (I think!) but can we make the cursor | visible inside the citation node?

Reply:
I agree that is really annoying but I don't see any easy fix. The issue is that when there is a reference node followed by the end of the citation node, there is no text node in between (because empty text nodes are not valid), so the cursor doesn't show